Avoid calling Page.getBlocks when relatively expensive

haozhun opened the pull request:
Avoid calling Page.getBlocks when each invocation does not correspond to other
operations that are at least on the order of Page.positionCount.

Perf profiling showed significant CPU time spent inside JVM_Clone. Perf
backtrace points to Page.getBlocks and MultiChannelGroupByHash. Perf also
showed significant CPU time spent inside TypeArrayKlass::allocate_common, which
is likely related.

cberner commented on presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorBucketFunction.java:
```diff
@@ -37,7 +37,8 @@ public RaptorBucketFunction(int bucketCount)
     {
         long hash = 0;
-        for (Block block : page.getBlocks()) {
+        for (int i = 0; i < page.getPositionCount(); i++) {
```
This doesn't look right. You should be iterating over channels, not positions

cberner commented:
You forgot to change the join compiler. If there aren't tests that would have caught that, make sure to add some

haozhun commented:
I have addressed your comments. Travis was able to catch both the `getChannelCount`/`getPagesCount` bug and the generated code that extends `PagesHashStrategy`. Therefore I didn't add additional tests.

haozhun commented:
@cberner, Travis passes after addressing your comments.

cberner commented:
Looks good

haozhun commented:
Merged #5312
